This fails the TCK with:

```
2025-11-18T16:35:18.9545649Z [ERROR] Failures: 
2025-11-18T16:35:18.9547688Z [ERROR]   JAXRSClient0138.getAcceptableMediaTypesIsSortedTest:195->invokeRequestAndCheckResponse:801->invokeRequestAndCheckResponse:812->JaxrsCommonClient.invoke:212 [WebValidatorBase] Unable to find the following search string in the server's response: 'text/html;q=0.7' at index: 17
2025-11-18T16:35:18.9549155Z [WebValidatorBase] Server's response:
2025-11-18T16:35:18.9549804Z -------------------------------------------
2025-11-18T16:35:18.9550268Z text/html;level=1 text/html */* text/html;level=2 text/* ;GETACCEPTABLEMEDIATYPES
2025-11-18T16:35:18.9550723Z -------------------------------------------
```

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/HeaderUtil.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/HeaderUtil.java
@@ -287,6 +287,7 @@ public class HeaderUtil {
                 list.add(MediaTypeHelper.valueOf(accept.trim()));
             }
         }
+        // Sort by weight (q) while preserving header-provided order on total ties
         MediaTypeHelper.sortByWeight(list);
         return list;
     }

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java
@@ -127,25 +127,6 @@ public class MediaTypeHelper {
             if (!isWildcardCompositeSubtype(mediaType.getSubtype()) && isWildcardCompositeSubtype(mediaType2.getSubtype()))
                 return 1;
 
-            int numNonQ = 0;
-            if (mediaType.getParameters() != null) {
-                numNonQ = mediaType.getParameters().size();
-                if (wasQ)
-                    numNonQ--;
-            }
-
-            int numNonQ2 = 0;
-            if (mediaType2.getParameters() != null) {
-                numNonQ2 = mediaType2.getParameters().size();
-                if (wasQ2)
-                    numNonQ2--;
-            }
-
-            if (numNonQ < numNonQ2)
-                return -1;
-            if (numNonQ > numNonQ2)
-                return 1;
-
             return 0;
         }
     }
@@ -165,7 +146,22 @@ public class MediaTypeHelper {
         if (hasAtMostOneItem(types)) {
             return;
         }
-        types.sort(Q_COMPARATOR);
+        // Stable sort: preserve original order when comparator returns 0
+        List<IndexedMediaType> indexed = new ArrayList<>(types.size());
+        for (int i = 0; i < types.size(); i++) {
+            indexed.add(new IndexedMediaType(types.get(i), i));
+        }
+        indexed.sort(new Comparator<IndexedMediaType>() {
+            @Override
+            public int compare(IndexedMediaType a, IndexedMediaType b) {
+                int cmp = Q_COMPARATOR.compare(a.mediaType, b.mediaType);
+                return cmp != 0 ? cmp : Integer.compare(a.originalIndex, b.originalIndex);
+            }
+        });
+        types.clear();
+        for (IndexedMediaType imt : indexed) {
+            types.add(imt.mediaType);
+        }
     }
 
     public static void sortByQSWeight(List<MediaType> types) {
@@ -364,5 +360,8 @@ public class MediaTypeHelper {
         }
 
         return count;
+    }
+
+    private record IndexedMediaType(MediaType mediaType, int originalIndex) {
     }
 }

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/resource/basic/AcceptHeaderOrderTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/resource/basic/AcceptHeaderOrderTest.java
@@ -1,0 +1,47 @@
+package org.jboss.resteasy.reactive.server.vertx.test.resource.basic;
+
+import static io.restassured.RestAssured.given;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.stream.Collectors;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.server.vertx.test.framework.ResteasyReactiveUnitTest;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class AcceptHeaderOrderTest {
+
+    @RegisterExtension
+    static ResteasyReactiveUnitTest test = new ResteasyReactiveUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Resource.class));
+
+    @Path("/headers")
+    public static class Resource {
+        @GET
+        @Path("/amt")
+        public String amt(@Context HttpHeaders hs) {
+            return hs.getAcceptableMediaTypes().stream().map(MediaType::toString).collect(Collectors.joining(","));
+        }
+    }
+
+    @Test
+    public void preservesOrderOnTies() {
+        String response = given()
+                .header("Accept",
+                        "application/json,text/html; charset=UTF-8,text/plain; charset=UTF-8,*/*;q=0.8")
+                .get("/headers/amt")
+                .then()
+                .statusCode(200)
+                .extract()
+                .asString();
+        assertEquals("application/json,text/html;charset=UTF-8,text/plain;charset=UTF-8,*/*;q=0.8", response);
+    }
+}

